Topic 2 design patterns and oop in javascript

davidleonel opened the pull request.

davidleonel committed:
Topic 2_ 1,2,3,4,5,6 NOT Completed

1-Create a Track class:

Track
- title : String
- artist : String
- duration : Number
+ play()
+ stop()
+ set(attr: String, value)
+ get(attr: String)

2-Instantiate some of your favorite tracks and play with them in the
console.

3-Create an Observable class which adds the logic to notify events to
their listeners.

Observable
- observers : Array
+ addObserver(event: String, observer: Object)
+ notify(event: String)
4-Create a TrackObserver class that listens for multiple events.

5-Publish "playing" event on Track.play(). You should be able to do
something like this in the console:

var droid = new Track();
var trackObserver = new TrackObserver([...]);

droid.set('artist', 'Jordan Suckley');
droid.set('title', 'Droid');
droid.addObserver(trackObserver);

droid.play(); // console: Playing Droid...
6-Publish "stopped" event on Track.stop().

davidleonel committed:
Topic 2: Minor changes

Done up to 7 -- Log to console when each event is fired.

davidleonel committed:
Topic 2 -- 8

8-Refactor Track class as a Module keeping your previous code for
reference.
DONE

davidleonel committed:
Topic 2: 8

Added a few changes, and comments. Also added the downloadableTrack
class

davidleonel committed:
Topic 2: Added comments

Added comments

troyax commented on 02-oop-inheritance/index.js:
```diff
@@ -0,0 +1,23 @@
+
+// I dont understand why i have to set the correct prototype to DownloadableTrack.
+//doing this works, but may be i didnt understand what it is that the exercise is asking for :S haha
+$.extend(TrackModule, DownloadableTrack);
```
Try write yourself the code of the extend, do not use jquery right now. The idea is to understand and know how to extend an object here :)

troyax commented on 02-oop-inheritance/index.js:
```diff
@@ -0,0 +1,23 @@
+ //big file, le tme know if you want me to change it and ill put 
+ //everything in just one single file
+
+(function (){
```
Is not necessary to make a module for the index.js file

troyax commented on 02-oop-inheritance/track.js:
```diff
@@ -0,0 +1,64 @@
+				this[temp] = value;
+			},
+			get : function(attr){
+				var temp = attr;	
```
The init method will not work, you are setting the 3 variables as private not allowing you to reach them. You might need to refactor this.

troyax commented on 02-oop-inheritance/trackObserver.js:
```diff
@@ -0,0 +1,5 @@
+//Constructor Function
+TrackObserver  = function(event){};
```
What do an observer has ? It should have multiple event -> callback pairs. If you need some explanation please let me know

davidleonel committed:
Topic 2: Fixed mistakes.

Fixed mistakes.

davidleonel committed:
Topic 2 ,8,9,10,11,12,13 DONE

Refactor Track class as a Module keeping your previous code for
reference.

Create a DownloadableTrack class that extends from Track adding a
download method. Here you will have to set the correct prototype to
DownloadableTrack.

Create a mixin object called Social with the methods: share(friendName)
and like().

Apply the mixin to Track class and play with the console output. You
should be able to do something like this in the console:

var droid = new Track();

droid.set('title', 'Droid');

// ...
droid.share('V. Rivas'); // output: Sharing Droid with V. Rivas
Create an Artist class and create some artists from one of your favorite
tracks.

Show how you would add an array of artists to a Track object.

Topic 2 Completed.

troyax commented:
Just one question, how are you testing that this is currently working? :P

troyax commented on 02-oop-inheritance/script.js:
```diff
@@ -0,0 +1,150 @@
+            
+            if (this.observers[i].eventCallback[j] === event) {
+                if (event === 'play') {
+                    this.observers[i].play(this);
```
you could do something to avoid the ifs, try to use the event as key to get the right method to call.

troyax commented on 02-oop-inheritance/script.js:
```diff
@@ -0,0 +1,150 @@
+};
+
+TrackObserver.prototype.play = function (track) {
+    console.log('Playing: ' + track.title + '.');
```
Why is the play, stop and download methods into the observer?

troyax commented on 02-oop-inheritance/script.js:
```diff
@@ -0,0 +1,150 @@
+//----------------------------TRACK----------------------
+var Track = (function () {
+
+	function Track (title, artist, duration, artists){
```
This is better :D

davidleonel committed:
Topic 2: Fixed issues.

Only one little thing left. i explained in the comments under the
observer class.

davidleonel committed:
Topic 2: Fixed mistakes.

Fixed the issue that i had, not being able to display the track´s name.
now the tracks name shows up on the corresponding events :)

davidleonel committed:
Topic 2: Final changes.

Moved the code around because some methods were not working do to bad
positioning from inside of the script.

troyax commented:
Now this PR seems to be DONE :D, just one thing, try to write functions definitions as

```
var sarasa = function (...) { ... };
```

This will keep your code a little bit cleaner
